leetcode/easy/958._Check_Completeness_of_a_Binary_Tree.py

```diff
@@ -0,0 +1,209 @@
+                                   root
+                                left  right
+                                x  x
+                                '''
+                                return UpDownNode(left_depth)
+                            else:
+                                '''
+                                     root
+                                  left  right
+                                 x   x
+                                x x x x
+
+                                (or flipped)
+                                '''
+                                return None
+                        case UpDownNode(right_depth):
+                            if left_depth == right_depth:
+                                '''
+                                    root
+                                left  right
+                                x  x  x
+                                '''
+                                return UpDownNode(right_depth)
+                            else:
+                                '''
+                                     root
+                                  left  right
+                                 x   x  x
+                                x x x x
+                                '''
+                                return None
+                case UpDownNode(left_depth):
+                    if root.right is None:
+                        '''
+                            root
+                        left
+                        x
+                        '''
+                        return None
+                    match isCompleteTreeRec(root.right, next_depth):
+                        case None:
+                            return None
+                        case FlatNode(right_depth):
+                            '''
+                                root
+                            left   right
+                            x
+                            '''
+                            if left_depth == right_depth + 1:
+                                return UpDownNode(left_depth)
+                            '''
+                                root
+                            left   right
+                            x      x   x
+                            '''
+                            return None
+                        case UpDownNode(_):
+                            '''
+                                root
+                            left   right
+                            x     x
+                            '''
+                            return None
+
+        return isCompleteTreeRec(root, 0) is not None
+
+'''
+BST:
+
+This is basically a spin-off of the level-order traversal.
+If we do a BST, whenever we start to see None, there should only be none.
+'''
+class Solution:
+    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
+        waiting_none = True
+
+        queue = [root]
+        while queue:
+            current_node = queue.pop(0)
```

Comment:
This `.pop(0)` is a very expensive operation O(n). If you want a true queue, you need to use Python's `collections.deque`